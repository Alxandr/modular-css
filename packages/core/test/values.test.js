"use strict";

const dedent = require("dedent");
const namer = require("test-utils/namer.js");
const { find, cwd } = require("test-utils/fixtures.js");

const Processor = require("../processor.js");

describe("/processor.js", () => {
    describe("values", () => {
        var processor;
        
        beforeEach(() => {
            processor = new Processor({
                cwd,
                namer,
            });
        });
        
        it("should fail on invalid value syntax", async () => {
            await expect(processor.string(
                "./invalid/value.css",
                "@value foo, bar from nowhere.css"
            )).rejects.toMatchObject({
                message : expect.stringContaining(`SyntaxError: Expected source but "n" found.`),
            });
        });

        it("should fail if a value imports a non-existant reference", async () => {
            await expect(processor.string(
                "./invalid/value.css",
                "@value not-real from \"../local.css\";"
            )).rejects.toMatchObject({
                message : expect.stringContaining(`Unable to locate "../local.css" from `),
            });
        });

        it("should support simple values", async () => {
            await processor.string(
                find("values.css"),
                dedent(`
                    @value a: red;
                    @value b:
                        Segoe UI
                        sans-serif;

                    .a {
                        color: a;
                        font-family: b;
                    }
                `)
            );
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });

        it("should support local values in value composition", async () => {
            const result = await processor.string(
                find("simple.css"),
                dedent(`
                    @value o: one;
                    @value local: './local.css';
                    @value o from local;
                    .fooga { background: one; }
                `)
            );

            expect(result.exports).toMatchSnapshot();
        });

        it("should support importing variables from a file", async () => {
            await processor.file(find("value-import.css"));
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });

        it("should support exporting imported variables", async () => {
            await processor.file(find("value-export.css"));
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });

        it("should support value composition", async () => {
            await processor.file(find("value-composition.css"));
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });

        it("should support value namespaces", async () => {
            await processor.file(find("value-namespace.css"));
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });

        it("should support value replacement in :external(...)", async () => {
            await processor.file(find("externals.css"));
            
            const result = await processor.output();
            
            expect(result.css).toMatchSnapshot();
        });
    });
});
