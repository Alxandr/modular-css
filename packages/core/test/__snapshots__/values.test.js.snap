// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js values should support exporting imported variables 1`] = `
"/* values.css */

/* value-import.css */
.red {
    color: red;
}

.green {
    color: green;
}

/* value-export.css */
.red {
    color: red;
}

.green {
    color: green;
}
"
`;

exports[`/processor.js values should support importing variables from a file 1`] = `
"/* values.css */
/* value-import.css */
.red {
    color: red;
}

.green {
    color: green;
}
"
`;

exports[`/processor.js values should support local values in value composition 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "local": "'./local.css'",
  "o": "red",
  "one": "red",
}
`;

exports[`/processor.js values should support simple values 1`] = `
"/* values.css */
.a {
    color: red;
    font-family: Segoe UI
    sans-serif;
}"
`;

exports[`/processor.js values should support value composition 1`] = `
"/* values.css */
/* value-composition.css */
.red {
    color: red;
    background: blue;
}
"
`;

exports[`/processor.js values should support value namespaces 1`] = `
"/* values.css */
/* value-namespace.css */
.red {
    color: red;
}

.blue {
    color: blue;
}

.other {
    color: #000;
}
"
`;

exports[`/processor.js values should support value replacement in :external(...) 1`] = `
"/* folder/folder.css */
.folder { margin: 2px; }
/* simple.css */
.fooga {
    color: red;
}
/* local.css */
.booga { background: green; }
/* externals.css */
.a .fooga {
    color: green;
}
.b:hover .booga {
    background: blue;
}
.fooga .booga {
    color: red;
}
"
`;
